### ISSUE TITLE
How to teardown database that was initialised in my code?

### ISSUE THREAD

**Issue by tomitrescak** [wallabyjs/public#2244]
### Issue description or question

I am trying to close any db connection that was created in my code.
I put the reference to it to "global.db = myDbObject".
I cannot access the "global" scope from the code in the teardown function.

Is there any nice trick how to achieve this?

**Comment by ArtemGovorov** [comment]
(assuming that you're using Jest) Jest creates its own execution scopes, so globals are not shared with Wallaby teardown function. You may use [Jest specific way](https://stackoverflow.com/questions/54488148/jest-global-teardown-runs-before-tests-finish) to execute some code after all of your tests.

**Comment by tomitrescak** [comment]
Still no luck:

I do this in my file:

```
export const photon = new Prisma.Photon();
global.photon = photon;
```

and this in teradown.ts

```
  console.log('!!!!!');
  console.log(global.photon);
```

This is what I see in wallaby console

```
console.log: !!!!!
console.log: undefined
```

This is my jest.config.js

```
module.exports = {
  modulePaths: ['<rootDir>/src/'],
  globalTeardown: '<rootDir>/src/tests/teardown.ts'
};
```

**Comment by ArtemGovorov** [comment]
The issue is not Wallaby specific, it is expected and [described in Jest docs](https://jestjs.io/docs/en/configuration.html#globalsetup-string). The referenced [StackOverflow answer](https://stackoverflow.com/a/54488526/146850) explains why it happens and how to deal with it.

**Comment by tomitrescak** [comment]
@ArtemGovorov there seems to be a Wallaby issue as all works in jest for me :/
Looks like wallaby refuses to execute the file as it is a typescript. So I am getting this mesage:

```
import { photon } from '../config/db';
[Info]         ^
[Info]  SyntaxError: Unexpected token {
```

**Comment by ArtemGovorov** [comment]
This looks like a different issue though. Could you please set up a small sample repo demonstrating the issue? We are happy to investigate.

**Comment by tomitrescak** [comment]
[error.zip](https://github.com/wallabyjs/public/files/3749589/error.zip)

Here you go ... just do `yarn`
Running `yarn jest` works fine
Running wallaby and in console you'll see the error

[Info]  import { bar } from './foo';
[Info]         ^
[Info]  SyntaxError: Unexpected token {

**Comment by tomitrescak** [comment]
If you prefer to set it up on github I can ...

**Comment by ArtemGovorov** [comment]
Thanks, we'll take a look tomorrow. However, after taking a quick look at the project, looks like you are using `globalTeardown` and because of how Jest works, it will not be able to access global variables set up in your tests anyway. Have you tried using `setupFiles` + `afterAll` as [described here](https://stackoverflow.com/questions/54488148/jest-global-teardown-runs-before-tests-finish/54488526#54488526)?

**Comment by tomitrescak** [comment]
@ArtemGovorov this leads to `[Error] Runtime error: ReferenceError: afterAll is not defined`

**Comment by tomitrescak** [comment]
Even if we set it up, will it "ALWAYS" execute? Wallaby tends to not execute it, if you keep modifying the text and kills the process without cleaning up ... ate least that was the reasoning before to use the "teardown" in wallaby setup.

**Comment by ArtemGovorov** [comment]
> Even if we set it up, will it "ALWAYS" execute?

Unfortunately no testing framework can guarantee `teardown` code execution. For example, if you by accident introduce say an infinite loop in Jest CLI watch mode , `teardown`/`afterAll` can not be executed in the same process.

**Comment by tomitrescak** [comment]
@ArtemGovorov the code above works with `setupFilesAfterEnv`.

Unfortunately Wallaby is killing the jest process as I type and `afterAll` is not executed. As a result, few seconds later I receive `maxConnections` exception and I can see bunch of hanging connections there :/

**Comment by ArtemGovorov** [comment]
You may also try the following (never tried it, so not sure if it works or not):

- in your Jest setup file, put a reference to your db object to `global.$_$tracer` (it is Wallaby specific): `global.$_$tracer.db = ...`

- now you can try accessing `global.$_$tracer.db` from Wallaby `setup` function.

You'll need to add Wallaby config like this (and switch from auto-config to this config file in VS Code via `Select Config` command):

```
module.exports = () => ({
   autoDetect: true,
   workers: {
       initial: 1,
       regular: 1
     },
   setup: () => {
     if(global.$_$tracer.db) global.$_$tracer.db.destroyAllConnections();
   }
});
```

If the approach works, you'll be able to make sure that all connections are destroyed before tests are executed.

**Comment by tomitrescak** [comment]
Yah, I had it there from the start. Must be package specific thing.
Currently I got a very hacky solutionm where I add the `db` instance to `fs` package:

 ```
   setup: function(wallaby) {
      var fs = require('fs');
      if (fs.db) {
        console.log('We still got it!!!');
        return fs.db.connect();
      }
      var Prisma = require('@generated/photon');
      var photon = new Prisma.Photon();
      fs.db = photon;
    },
    teardown: function(wallaby) {
      var fs = require('fs');
      return fs.db.disconnect();
    }
```

weird bit for me is that the `db` survives between run in fs package.

**Comment by tomitrescak** [comment]
Actually, I could do it inside my code and use auto-generated config. This will NOT clean up though:

```js
import * as Prisma from '@generated/photon';
import fs from 'fs';

const wfs: any = fs;
export let photon: Prisma.Photon;

if (wfs.db) {
  console.log('We still got it!!!');
  photon = wfs.db;
} else {
  console.log('New photon');
  photon = new Prisma.Photon();
  wfs.db = photon;
}
```

**Comment by smcenlly** [comment]
Hi @tomitrescak - as you suggested in a previous post, Wallaby wasn't correctly executing the file because it was TypeScript. 

We've updated our Jest integration to cater for this scenario and Wallaby should now work exactly the same as Jest. This is available in the latest version of Wallaby core, v1.0.769. If Wallaby does not update for you automatically, you may [force a core update](https://wallabyjs.com/docs/intro/troubleshooting.html#forcing-wallabyjs-core-update).
